src/Keboola/StorageApi/Client.php

```diff
@@ -1434,13 +1515,127 @@ public function uploadSlicedFile(array $slices, FileUploadOptions $options, File
             $manifestUploadOptions['ServerSideEncryption'] = $uploadParams['x-amz-server-side-encryption'];
         }
         $s3Client->putObject($manifestUploadOptions);
+    }
 
-        // Cleanup
-        if ($fs) {
-            $fs->remove($currentUploadDir);
+    public function downloadFile($fileId, $destionation, GetFileOptions $getOptions = null)
+    {
+        $fileInfo = $this->getFile($fileId, $getOptions);
+        switch ($fileInfo['provider']) {
+            case 'azure':
+                $this->downloadAbsFile($fileInfo, $destionation);
+                break;
+            default:
+                $this->logger->warning('File upload without provider has been used', $fileInfo);
+            case 'aws':
+                $this->downloadS3File($fileInfo, $destionation);
+                break;
         }
+    }
 
-        return $preparedFileResult['id'];
+    private function downloadAbsFile(array $fileInfo, $destination)
+    {
+        $blobClient = BlobRestProxy::createBlobService(
+            $fileInfo['absCredentials']['SASConnectionString']
+        );
+        $getResult = $blobClient->getBlob($fileInfo['absPath']['container'], $fileInfo['absPath']['name']);
+        file_put_contents($destination, $getResult->getContentStream());
+    }
+
+    private function downloadS3File(array $fileInfo, $destination)
+    {
+        $s3Client = new S3Client([
+            'version' => 'latest',
+            'region' => $fileInfo['region'],
+            'credentials' => [
+                'key' => $fileInfo['credentials']['AccessKeyId'],
+                'secret' => $fileInfo['credentials']['SecretAccessKey'],
+                'token' => $fileInfo['credentials']['SessionToken'],
+            ],
+        ]);
+        $object = $s3Client->getObject([
+            'Bucket' => $fileInfo['s3Path']['bucket'],
+            'Key' => $fileInfo['s3Path']['key'],
+        ]);
+        file_put_contents($destination, $object['Body']);
```

Comment:
Nenačte to celé do paměti při téhle implementaci? Radši bych udělal copy paste tady https://github.com/keboola/input-mapping/blob/5813b47f2948efec289a6ee2b8ef801e78d5b818/Reader/Reader.php#L196

Reply:
Nevěděl jsem, že to má "SaveAs" :+1:  to bude určitě lepší. Díky.